### PR TITLE
SWC-6608: define page size options that are aware of the data maxRows…

### PR DESCRIPTION
**Pull request by jay-hodgson**
…PerPage

### CODE REVIEW

**Comment by jay-hodgson** [comment]
https://github.com/Sage-Bionetworks/synapse-web-monorepo/assets/1864447/9497e0ff-98ab-421e-a816-d66a8710cfa7

File: packages/synapse-react-client/src/components/SynapseTable/TablePagination.tsx [hunk 1]
```diff
@@ -22,15 +34,16 @@ export const TablePagination = () => {
 
   // PORTALS-2259: Special case.  If we're on the first page,
   // and the total query count is less than the min page size, then do not show pagination UI.
-  if (currentPage == 1 && queryCount && queryCount < 10) {
+  // Also hide pagination if the query count is unavailable.
+  if ((currentPage == 1 && queryCount && queryCount < 10) || !queryCount) {
     return <></>
   }
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/SynapseTable/TablePagination.tsx:40]: I think 10 here should be pageSizeOptionsBasedOnData[0], right?
```suggestion
  const smallestPageSizeOption = pageSizeOptionsBasedOnData[0]
  if ((currentPage == 1 && queryCount && queryCount < smallestPageSizeOption) || !queryCount) {
    return <></>
  }
```

File: packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  })
+
+  it('renders pagination and page size elements', () => {
+    render(
+      Wrapper({
+        goToPage: mockGoToPage,
+        setPageSize: mockSetPageSize,
+        pageSize: 10,
+        currentPage: 1,
+      }),
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx:52]: Use JSX, not function call
```suggestion
      <Wrapper 
        goToPage={mockGoToPage}
        setPageSize={mockSetPageSize}
        pageSize={10}
        currentPage={1}
      } />,
```

File: packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    )
+    act(() => {
+      screen.findByText(renderedTextConfirmation)
+    })
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx:56]: You will also need to add `async` to the function definition
```suggestion
    await screen.findByText(renderedTextConfirmation)
```

File: packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+      setTableData({
+        concreteType: 'org.sagebionetworks.repo.model.table.QueryResultBundle',
+        queryCount: 100,
+        maxRowsPerPage: 100,
+      })
+    })
+
+    // Simulate changing the page
+    act(() => {
+      fireEvent.click(screen.getByText('2')) // Click on page 2
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx:92]: can you use `await userEvent.click` instead? You can remove the `act` if you do that

File: packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        maxRowsPerPage: 1000,
+      })
+    })
+
+    // Simulate changing the page size
+    act(() => {
+      fireEvent.change(screen.getByRole('combobox'), {
+        target: { value: '25' },
+      }) // Change page size to 25
+    })
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx:124]: If you can't do this one with `userEvent`, then you might need to wrap the following `expect` assertions in a `waitFor`

File: packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+    expect(mockSetPageSize).toHaveBeenCalledTimes(1)
+    expect(mockSetPageSize).toHaveBeenCalledWith('25')
+  })
+
+  it('hides pagination when on the first page and query count is less than 10', () => {
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx:130]: Consider adding one case where `maxRowsPerPage` is 5 and there are > 10 results--verify that pagination controls are visible and the combobox has one option that is 5

File: packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx (hunk 4 above)
Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx:92]: Sure thing, this was code modified from a chatgpt response :D

File: packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx (hunk 5 above)
Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx]: Replaced with userEvent

File: packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx (hunk 6 above)
Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/SynapseTable/TablePagination.test.tsx]: Done

File: packages/synapse-react-client/src/components/SynapseTable/TablePagination.tsx (hunk 1 above)
Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/SynapseTable/TablePagination.tsx]: This is an interesting case!  I think we can hide the pagination as soon as all of the results are shown in a single page

Comment (nickgros) [reply packages/synapse-react-client/src/components/SynapseTable/TablePagination.tsx]: I think that makes this unusual sequence possible:

1. Receive 40 results, 25 results shown per page
2. Change the results shown per page to 40
3. Pagination is now hidden and you can't change it to 10 or back to 25

Comment (jay-hodgson) [reply packages/synapse-react-client/src/components/SynapseTable/TablePagination.tsx]: This is true, so it works like the Show More button (expand to all results, no way to contract).  It's kind of strange, but I think the UX is ok.  If you disagree, let's talk about the scenario with design.  (How important is it to hide pagination when the results are less than the page size?)

**Review by nickgros — APPROVED** [review]

**Review by nickgros — APPROVED** [review]

**Pull request merged**